### PR TITLE
[yang-models] Add missing constraints in VLAN yang model. Fixes #12256

### LINKED ISSUE #12256: [yang-models] Missing constraints in VLAN, VLAN_INTERFACE yang models
#### Description

VLAN, VLAN_INTERFACE yang models are missing validation criteria. These constraints are missing from the YANG model, but they are checked in sonic-utilities using adhoc validation. Included in the description for each missing constraint is the adhoc validation code in sonic-utilities that we aim to replace with systemized YANG validation.

1. Don't allow write of default vlan (vid = 1)

[sonic-utilities config/vlan.py](https://github.com/sonic-net/sonic-utilities/blob/master/config/vlan.py)
```
        if vid == 1:
            ctx.fail("{} is default VLAN".format(vlan)) # TODO: MISSING CONSTRAINT IN YANG MODEL
```
2. If vlan id already exists, do not allow duplicitous write to ConfigDB

[sonic-utilities config/vlan.py](https://github.com/sonic-net/sonic-utilities/blob/master/config/vlan.py)
```
        if clicommon.check_if_vlanid_exist(db.cfgdb, vlan): # TODO: MISSING CONSTRAINT IN YANG MODEL
            ctx.fail("{} already exists".format(vlan))
```
3. Do not allow removal of vlan interface if vlan has IP address assigned

[sonic-utilities config/vlan.py](https://github.com/sonic-net/sonic-utilities/blob/master/config/vlan.py)
```
        for intf_key in intf_table:
            if ((type(intf_key) is str and intf_key == 'Vlan{}'.format(vid)) or # TODO: MISSING CONSTRAINT IN YANG MODEL
                (type(intf_key) is tuple and intf_key[0] == 'Vlan{}'.format(vid))):
                ctx.fail("{} can not be removed. First remove IP addresses assigned to this VLAN".format(vlan))
```

4. Don't allow removal of vlan if there are members assigned

[sonic-utilities config/vlan.py](https://github.com/sonic-net/sonic-utilities/blob/master/config/vlan.py)
```
        keys = [ (k, v) for k, v in db.cfgdb.get_table('VLAN_MEMBER') if k == 'Vlan{}'.format(vid) ]
    
        if keys: # TODO: MISSING CONSTRAINT IN YANG MODEL
            ctx.fail("VLAN ID {} can not be removed. First remove all members assigned to this VLAN.".format(vid))
```

5. If interface naming mode is alias, make sure corresponding port name exists

[sonic-utilities config/vlan.py](https://github.com/sonic-net/sonic-utilities/blob/master/config/vlan.py)
```
        if clicommon.get_interface_naming_mode() == "alias": # TODO: MISSING CONSTRAINT IN YANG MODEL
            alias = port
            iface_alias_converter = clicommon.InterfaceAliasConverter(db)
            port = iface_alias_converter.alias_to_name(alias)
            if port is None:
                ctx.fail("cannot find port name for alias {}".format(alias))

```

6.  Make sure port is not configured as a mirror destination port

[sonic-utilities config/vlan.py](https://github.com/sonic-net/sonic-utilities/blob/master/config/vlan.py)
```
        if clicommon.is_port_mirror_dst_port(db.cfgdb, port): # TODO: MISSING CONSTRAINT IN YANG MODEL
            ctx.fail("{} is configured as mirror destination port".format(port))
```

7. Make sure port is not already a vlan member

[sonic-utilities config/vlan.py](https://github.com/sonic-net/sonic-utilities/blob/master/config/vlan.py)
```
        if clicommon.is_port_vlan_member(db.cfgdb, port, vlan): # TODO: MISSING CONSTRAINT IN YANG MODEL
            ctx.fail("{} is already a member of {}".format(port, vlan))
```

8. Make sure port is not a router interface

[sonic-utilities config/vlan.py](https://github.com/sonic-net/sonic-utilities/blob/master/config/vlan.py)
```
        if (is_port and clicommon.is_port_router_interface(db.cfgdb, port)) or \
           (not is_port and clicommon.is_pc_router_interface(db.cfgdb, port)): # TODO: MISSING CONSTRAINT IN YANG MODEL
            ctx.fail("{} is a router interface!".format(port))
```

9. Make sure port is not already a portchannel member

[sonic-utilities config/vlan.py](https://github.com/sonic-net/sonic-utilities/blob/master/config/vlan.py)
```
        if (is_port and clicommon.interface_is_in_portchannel(portchannel_member_table, port)): # TODO: MISSING CONSTRAINT IN YANG MODEL
            ctx.fail("{} is part of portchannel!".format(port))
```

10. Make sure port is not already an untagged member

[sonic-utilities config/vlan.py](https://github.com/sonic-net/sonic-utilities/blob/master/config/vlan.py)
```
        if (clicommon.interface_is_untagged_member(db.cfgdb, port) and untagged): # TODO: MISSING CONSTRAINT IN YANG MODEL
            ctx.fail("{} is already untagged member!".format(port))
```
#### Steps to reproduce the issue:
1. Using YANG model as mode of validation, write to ConfigDB with an invalid configuration
2. Write to ConfigDB will go through even though it should be prohibited by YANG model
#### Describe the results you received:
Write to ConfigDB succeeds even though the configuration is invalid

#### Describe the results you expected:
Write to ConfigDB should fail

### PR DESCRIPTION
**Pull request by matiAlfaro**
Fix #12256 

Implement constraints as described in https://github.com/sonic-net/sonic-buildimage/issues/12256 

1. **Don't allow write of default vlan (vid = 1)**
- Change range from 1..4094 to 2..4094

2. **If vlan id already exists, do not allow duplicitous write to ConfigDB**
No change is needed, name is key so it will be unique

3. **Do not allow removal of vlan interface if vlan has IP address assigned**
No chnage is needed since both VLAN_INTERFACE_LIST and VLAN_INTERFACE_IPPREFIX_LIST reference /vlan:sonic-vlan/vlan:VLAN/vlan:VLAN_LIST/vlan:name
- 	Added IP_PREFIX_WITHOUT_CREATNG_VLAN test

4. **Don't allow removal of vlan if there are members assigned**
No chnage is needed since VLAN_MEMBER_LIST refereces "/vlan:sonic-vlan/vlan:VLAN/vlan:VLAN_LIST/vlan:name"
- Added test VLAN_MEMBERS_WITHOUT_CREATING_VLAN

5. **If interface naming mode is alias, make sure corresponding port name exists**
sonic-yang doesn't support adding a ports using alias

6. **Make sure port is not configured as a mirror destination port**
- must "not(/sms:sonic-mirror-session/sms:MIRROR_SESSION/sms:MIRROR_SESSION_LIST[sms:dst_port=current()])" 
- Fixed MIRROR_SESSION in sonic-yang-models\tests\files\sample_config_db.json (Ethernet2 is in Vlan111)
- Test MIRROR_SESSION_ON_VLAN_MEMBER_PORT

7. **Make sure port is not already a vlan member**
No change is needed, <vlan,port> is the key for VLAN_MEMBER_LIST

8. **Make sure port is not a router interface**
- must "not(/intf:sonic-interface/intf:INTERFACE/intf:INTERFACE_LIST[intf:name=current()])"
- Test VLAN_ADD_PORT_THAT_IS_ROUTER_INTERFACE

9. **Make sure port is not already a portchannel member**
- must "not(/lag:sonic-portchannel/lag:PORTCHANNEL_MEMBER/lag:PORTCHANNEL_MEMBER_LIST[lag:port=current()]/lag:name)"
- Test VLAN_ADD_PORT_CHANNEL_MEMBER

10. **Make sure port is not already an untagged member**
- must "not(/vlan:sonic-vlan/vlan:VLAN_MEMBER/vlan:VLAN_MEMBER_LIST[vlan:port=current() and vlan:tagging_mode='untagged' and vlan:name!=current()/../name])"
- Test VLAN_ADD_PORT_THAT_IS_UNTAGGED

**Verify Vlan name matches vlanid**
- must "substring-after(../name, 'Vlan') = current()"
- Test VLAN_CREATE_VLAN_WITH_MISSMATCHING_NAME
##### Work item tracking
- Microsoft ADO **(number only)**:

#### How to verify it
Added tests to verify all modified and added constraints.

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205
- [ ] 202211
- [ ] 202305

#### Tested branch (Please provide the tested image version)

- [ ] 
- [ ]

### CODE REVIEW

**Comment by matiAlfaro** [comment]
@isabelmsft - this should fix https://github.com/sonic-net/sonic-buildimage/issues/12256

File: src/sonic-yang-models/yang-models/sonic-vlan.yang [hunk 1]
```diff
@@ -178,14 +189,24 @@ module sonic-vlan {
...
+					must "not(current() = 'Vlan1')"
+					{
+						error-message "Vlan1 is default Vlan";
+					}
+				} 						
 				leaf vlanid {
 					type uint16 {
 						range 1..4094;
 					}
+					must "not(current() = 1)"
```
Comment (ganglyu) [inline src/sonic-yang-models/yang-models/sonic-vlan.yang:203]: Can we change range to 2..4094?

File: src/sonic-yang-models/yang-models/sonic-vlan.yang [hunk 2]
```diff
@@ -250,6 +270,19 @@ module sonic-vlan {
...
+					must "not(/sms:sonic-mirror-session/sms:MIRROR_SESSION/sms:MIRROR_SESSION_LIST[sms:dst_port=current()])" 
+					{
+    					error-message "Port is used as a destination port in a mirror session.";
+					}
+					must "not(/lag:sonic-portchannel/lag:PORTCHANNEL_MEMBER/lag:PORTCHANNEL_MEMBER_LIST[lag:port=current()]/lag:name)" {
+    					error-message "Port is a member of a port channel.";
+					}
+					must "not(/intf:sonic-interface/intf:INTERFACE/intf:INTERFACE_LIST[intf:name=current()])" {
+                		error-message "Port is a router interface.";
+            		}
```
Comment (ganglyu) [inline src/sonic-yang-models/yang-models/sonic-vlan.yang:282]: Please use tab to replace space.

File: src/sonic-yang-models/yang-models/sonic-vlan.yang (hunk 1 above)
Comment (matiAlfaro) [reply src/sonic-yang-models/yang-models/sonic-vlan.yang]: Yes, it does make more sense.

File: src/sonic-yang-models/yang-models/sonic-vlan.yang (hunk 2 above)
Comment (matiAlfaro) [reply src/sonic-yang-models/yang-models/sonic-vlan.yang]: done.

**Review by isabelmsft — APPROVED** [review]

**Review by ganglyu — APPROVED** [review]

**Pull request merged**